addon-test-support/-private/mock/proxy.js

```diff
@@ -1,39 +1,113 @@
+import { assert } from '@ember/debug';
 import mockFunction from './function';
 
-export default function proxyFactory(original) {
-  let holder = {};
+const PROXIES = new WeakMap();
 
-  let proxy = new Proxy(original, {
-    get (target, name) {
-      if (name === '_reset') {
-        return () => holder = {};
+function getProxyConfig(proxy) {
+  assert(`Could not find '${proxy}' in the proxy map.`, PROXIES.has(proxy));
+  return PROXIES.get(proxy);
+}
+
+function assertConfigurableDescriptor(target, key) {
+  const descriptor = Object.getOwnPropertyDescriptor(target, key);
+  assert(
+    `Cannot override non-configurable descriptor for '${key}' on '${target}'.`,
+    !descriptor || descriptor.configurable
+  );
+}
+
+function assertWritableDescriptor(target, key) {
+  const descriptor = Object.getOwnPropertyDescriptor(target, key);
+  if (descriptor && 'configurable' in descriptor && !descriptor.configurable) {
+    assert(
+      `Cannot directly set '${key}' on '${target}', because it is a non-configurable and non-writable property.\nSee https://github.com/kaliber5/ember-window-mock/issues/99`,
+      typeof descriptor.set === 'function' || descriptor.writable
+    );
+  }
+}
+
+export default function proxyFactory(
+  original,
+  { makeHolder = () => ({}), makeDescriptors = () => ({}) } = {}
```

Comment:
Uh, tbh I just learned that you can have default values in a destructuring expression, somehow didn't encounter this before. Nice! :)